Sources/SourceKitLSP/Swift/SemanticTokens.swift

```diff
@@ -0,0 +1,352 @@
+
+  init(sourcekitd: SourceKitD, snapshot: DocumentSnapshot, useName: Bool = false) {
+    self.sourcekitd = sourcekitd
+    self.snapshot = snapshot
+    self.useName = useName
+  }
+
+  func parseTokens(_ response: SKDResponseDictionary) -> [SemanticToken] {
+    let keys = sourcekitd.keys
+    var tokens: [SemanticToken] = []
+
+    if let offset: Int = useName ? response[keys.nameoffset] : response[keys.offset],
+       let length: Int = useName ? response[keys.namelength] : response[keys.length],
+       let start: Position = snapshot.positionOf(utf8Offset: offset),
+       let skKind: sourcekitd_uid_t = response[keys.kind],
+       let (kind, modifiers) = parseKindAndModifiers(skKind) {
+      let token = SemanticToken(
+        start: start,
+        length: length,
+        kind: kind,
+        modifiers: modifiers
+      )
+      tokens.append(token)
+    }
+
+    if let substructure: SKDResponseArray = response[keys.substructure] {
+      tokens += parseTokens(substructure)
+    }
+
+    return tokens
+  }
+
+  func parseTokens(_ response: SKDResponseArray) -> [SemanticToken] {
+    var result: [SemanticToken] = []
+    response.forEach { (_, value) in
+      result += parseTokens(value)
+      return true
+    }
+    return result
+  }
+
+  private func parseKindAndModifiers(_ uid: sourcekitd_uid_t) -> (SemanticToken.Kind, SemanticToken.Modifiers)? {
+    let values = sourcekitd.values
+    switch uid {
+    case values.kind_keyword,
+         values.syntaxtype_keyword:
+      return (.keyword, [])
+    case values.syntaxtype_attribute_builtin:
+      return (.modifier, [])
+    case values.decl_module:
+      return (.namespace, [])
+    case values.decl_class:
+      return (.class, [.declaration])
+    case values.ref_class:
+      return (.class, [])
+    case values.decl_struct:
+      return (.struct, [.declaration])
+    case values.ref_struct:
+      return (.struct, [])
+    case values.decl_enum:
+      return (.enum, [.declaration])
+    case values.ref_enum:
+      return (.enum, [])
+    case values.decl_protocol:
+      return (.interface, [.declaration])
+    case values.ref_protocol:
+      return (.interface, [])
+    case values.decl_associatedtype,
+         values.decl_typealias,
+         values.decl_generic_type_param:
+      return (.typeParameter, [.declaration])
+    case values.ref_associatedtype,
+         values.ref_typealias,
+         values.ref_generic_type_param:
+      return (.typeParameter, [])
+    case values.decl_function_constructor,
+         values.decl_function_subscript,
+         values.decl_function_free:
+      return (.function, [.declaration])
+    case values.decl_function_method_static,
+         values.decl_function_method_instance,
+         values.decl_function_method_class:
+      return (.method, [.declaration])
+    case values.ref_function_constructor,
+         values.ref_function_destructor,
+         values.ref_function_free,
+         values.ref_function_subscript:
```

Comment:
Same discussion about constructors and subscripts here. And destructors are IMHO normal (non-static) methods.